src/backend/backend.h

```diff
@@ -89,41 +108,92 @@ namespace snmalloc
 
       CapPtr<void, CBChunk> p;
-      if (local_state != nullptr)
+      if ((local_state != nullptr) && (size <= MAX_CACHED_SIZE))
       {
-        p =
-          local_state->local_address_space.template reserve_with_left_over<PAL>(
-            size);
+#ifdef SNMALLOC_CHECK_CLIENT
+        auto& local = is_meta ? local_state->local_meta_address_space :
+                                local_state->local_address_space;
+#else
+        auto& local = local_state->local_address_space;
+#endif
+
+        p = local.template reserve_with_left_over<PAL>(size);
         if (p != nullptr)
-          local_state->local_address_space.template commit_block<PAL>(p, size);
-        else
         {
-          auto& a = h.address_space;
-          // TODO Improve heuristics and params
-          auto refill_size = bits::max(size, bits::one_at_bit(21));
-          auto refill = a.template reserve<false>(refill_size, h.pagemap);
-          if (refill == nullptr)
-            return nullptr;
-          local_state->local_address_space.template add_range<PAL>(
-            refill, refill_size);
-          // This should succeed
-          p = local_state->local_address_space
-                .template reserve_with_left_over<PAL>(size);
-          if (p != nullptr)
-            local_state->local_address_space.template commit_block<PAL>(
-              p, size);
+          return p;
         }
+
+        auto refill =
+          global.template reserve<false>(LOCAL_CACHE_BLOCK, h.pagemap);
+        if (refill == nullptr)
+          return nullptr;
+        auto refill_size = LOCAL_CACHE_BLOCK;
+
+#ifdef SNMALLOC_CHECK_CLIENT
+        if (is_meta)
+        {
+          constexpr size_t OFFSET_MASK =
+            LOCAL_CACHE_BLOCK - LOCAL_CACHE_META_BLOCK;
+          // Don't use first or last block in the larger reservation
+          // Loop required to get uniform distribution.
+          size_t offset;
+          do
+          {
+            offset = get_entropy64<PAL>() & OFFSET_MASK;
+          } while ((offset == 0) || (offset == OFFSET_MASK));
+
+          refill_size = LOCAL_CACHE_META_BLOCK;
+          refill = pointer_offset(refill, offset);
+        }
+#endif
+        local.template commit_block<PAL>(refill, refill_size);
+        local.template add_range<PAL>(refill, refill_size);
+
+        // This should succeed
+        return local.template reserve_with_left_over<PAL>(size);
       }
-      else
+
+#ifdef SNMALLOC_CHECK_CLIENT
+      // During start up we need meta-data before we have a local allocator
+      // This code protects that meta-data with randomisation, and guard pages.
+      if (local_state == nullptr && is_meta)
       {
-        auto& a = h.address_space;
-        p = a.template reserve_with_left_over<true>(size, h.pagemap);
+        size_t rsize = bits::max(OS_PAGE_SIZE, bits::next_pow2(size));
+        size_t size_request = rsize * 64;
+
+        p = global.template reserve<false>(size_request, h.pagemap);
```

Comment:
Would it make sense to create a method for the pattern of using a larger allocation from an address space to create a randomized inner object? It is only used in two places, but would make it easier to make changes uniformly later on.

Reply:
Done.